### ISSUE TITLE
Issue with Layer Freezing Despite ‘freeze=None’ in YOLOv8 Pose Training Configuration

### ISSUE THREAD

**Issue by 666echo** [ultralytics/ultralytics#9392]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

I am currently working on pose training using YOLOv8 and encountered an unexpected behavior where the layer 'model.22.dfl.conv.weight' is being frozen despite setting freeze=None. This is happening in my configuration where I specifically require all layers to be trainable to fine-tune the model on pose estimation tasks. Below is the relevant model layer information:

---------------------------------------------------------------------------
 22        [15, 18, 21]  1    954970  ultralytics.nn.modules.head.Pose             [1, [13, 3], [64, 128, 256]]  
YOLOv8_g_ghost_VoVGSCSP summary: 346 layers, 2668474 parameters, 2668458 gradients, 7.0 GFLOPs

Freezing layer 'model.22.dfl.conv.weight'
---------------------------------------------------------------------------

Despite the absence of freeze instructions (freeze=None), the layer 'model.22.dfl.conv.weight' still gets frozen, which is not the intended behavior for my pose training setup. I need all layers, especially this one, to be active for training to ensure the model is fully adapted to the pose estimation tasks.

Could you please provide guidance on why this layer is being frozen and how to ensure it remains trainable? Is there a specific configuration or code snippet I am missing that is necessary to prevent this layer from being frozen?

Thank you for your assistance.

**Comment by captainfffsama** [comment]
https://github.com/ultralytics/ultralytics/issues/9371

**Comment by 666echo** [comment]
> #9371

Thank you for your assistance.

**Comment by glenn-jocher** [comment]
@666echo hello! 👋 Thanks for reaching out with your question regarding the unexpected freezing of the 'model.22.dfl.conv.weight' layer during your pose training with YOLOv8.

This behavior indeed seems unintended when `freeze=None` is specified, as all layers should remain trainable in this case. To address the issue, let's ensure a couple of things:
- First, kindly verify that your training script or command line does not inadvertently set the `freeze` argument elsewhere.
- Next, double-check the configuration file (if you're using one) to ensure it doesn't specify layer freezing.

If everything seems correct and you're still facing this issue, you might be encountering an unexpected behavior or bug. I recommend updating to the latest version of our code base if not already done:

```bash
git pull
```

And if the issue persists, it could be very helpful to us if you could provide a minimal reproducible example including your command/script to run the training. That way, we can dive deeper into what might be causing this issue.

In the meantime, as a temporary workaround, you might explicitly set layers to trainable after model initialization in your script, though this approach is a bit manual:

```python
for name, param in model.named_parameters():
    if name == 'model.22.dfl.conv.weight':
        param.requires_grad = True
```

Please let us know if any of these steps helps or if the issue continues. We're here to help!

Thanks again for bringing this to our attention! 🚀

**Comment by 666echo** [comment]
@glenn-jocher Thank you so much for your prompt reply and assistance. Following your guidance, I revisited my training script and configuration file, and I indeed found that the issue was related to the setting . This specific setting was preventing the  layer from being unfrozen. Following your advice, I've successfully adjusted this setting, thereby unfreezing the  layer, and my model is now training as expected.

Your guidance was immensely helpful in quickly identifying and solving the problem. The issue was indeed due to the  setting, which caused the  layer to remain frozen unintentionally. After making the necessary adjustments based on your suggestions, the layer is now trainable, and I'm seeing positive progress in my model's training.

Thank you for your patience and valuable time. Your advice has not only helped me resolve the immediate issue but also provided me with further insights into model adjustment.

Thank you once again for your help!

**Comment by glenn-jocher** [comment]
@666echo You're very welcome! 😊 Delighted to hear that the guidance proved useful and you were able to resolve the issue by adjusting the specific setting. It's great to know your model is now training as expected and making positive progress. 

Remember, it’s all about those small adjustments that can sometimes make a big difference. If you have any more questions down the line or need further assistance, feel absolutely free to reach out. 

Keep up the fantastic work on your project, and happy training! 🚀
